3d visualize masks update mesh manager interface

hanslovsky opened the pull request:
This PR is for keeping track of changes more easily than through the CLI before merging back into the original feature branch

igorpisarev commented on src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/MeshManagerWithSingleMesh.kt:
```diff
@@ -89,4 +101,22 @@ class MeshManagerWithSingleMesh<Key>(
+    private fun setupGeneratorState(state: MeshGenerator.State) {
+        LOG.debug("Setting up state for mesh key {}", meshKey)
+        state.settings.finestScaleLevel = 1
+        state.settings.finestScaleLevel = 0
```
One of these should probably be the coarsest scale level? Although, since in the next lines it gets bound to `settings`, it seems that the values will be overwritten by the binding anyway

hanslovsky replied:
Thanks for catching this @igorpisarev

hanslovsky commented:
This will be ready to be merged soon, and subsequently #377 as well. I refactored the mesh managers quite a bit. After all, we do not need a `MeshManager` interface anymore because the sources are now responsible for handling events etc and they have full access to all their members, including the mesh managers.

 1. The bulk of the work happens in the [`AdaptiveResolutionMeshManager`](https://github.com/saalfeldlab/paintera/blob/faa0aea8940cc050ef57544e61eee9b713ab8955/src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/adaptive/AdaptiveResolutionMeshManager.kt), formerly known as `AbstractMeshManager` with a few adaptations.
 2. The two current mesh managers are [`MeshManagerWithSingleMesh`](https://github.com/saalfeldlab/paintera/blob/faa0aea8940cc050ef57544e61eee9b713ab8955/src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/MeshManagerWithSingleMesh.kt) (manages a single mesh for thresholded sources and similar) and [`MeshManagerWithAssignmentForSegments`](https://github.com/saalfeldlab/paintera/blob/faa0aea8940cc050ef57544e61eee9b713ab8955/src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/MeshManagerWithAssignmentForSegments.kt) (neuron data with fragment-segment assignments) do not implement any interfaces. They both have an `AdaptiveResolutionMeshManager` member and simply translate user requests into the appropriate format for that manager. This avoids inheritance and the `MeshManagerWithSingleMesh` can be used for pretty much any source that is a boolean mask.

As far as I can tell, most things work now. I am still testing a few things that I have not used as much while I was working on this feature. In particular, that is
 - [ ] de-/serialization of mesh settings
 - [ ] mesh generation/refreshing for painted canvases
 - [ ] mesh generation/refreshing after committing canvas

Besides this ongoing testing, I will also go through `AdaptiveResolutionMeshManager` with @igorpisarev to optimize that class and identify potential bugs/bad design.

hanslovsky commented:
Currently, this is not compiling with a cryptic kotlin error. This is the bisect:
```
1be0d52bd55851941b180001bb0367d88042d836 is the first bad commit
commit 1be0d52bd55851941b180001bb0367d88042d836
Author: Philipp Hanslovsky <hanslovskyp@janelia.hhmi.org>
Date:   Wed Jan 29 09:37:31 2020 -0500

    Simplify mesh management

 .../saalfeldlab/paintera/meshes/MeshExporter.java  |  14 +-
 .../saalfeldlab/paintera/meshes/MeshGenerator.java |  11 +-
 .../paintera/meshes/MeshGeneratorJobManager.java   |  11 +-
 .../paintera/state/IntersectingSourceState.java    |   9 +-
 .../paintera/state/ThresholdingSourceState.java    |   7 +-
 .../paintera/meshes/managed/GetBlockListFor.kt     |   7 +
 .../paintera/meshes/managed/GetMeshFor.kt          |  56 ++++++++
 .../paintera/meshes/managed/MeshManagerSettings.kt |  48 +++++++
 .../paintera/meshes/managed/MeshManagerStore.kt    |  15 +-
 .../MeshManagerWithAssignmentForSegments.kt        |   2 -
 .../meshes/managed/MeshManagerWithSingleMesh.kt    |   2 -
 .../adaptive/AdaptiveResolutionMeshManager.kt      | 152 +++++----------------
 12 files changed, 176 insertions(+), 158 deletions(-)
 create mode 100644 src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/GetBlockListFor.kt
 create mode 100644 src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/GetMeshFor.kt
 create mode 100644 src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/MeshManagerSettings.kt
```
```
$ git bisect log
git bisect start
# bad: [faa0aea8940cc050ef57544e61eee9b713ab8955] Fix formatting
git bisect bad faa0aea8940cc050ef57544e61eee9b713ab8955
# bad: [faa0aea8940cc050ef57544e61eee9b713ab8955] Fix formatting
git bisect bad faa0aea8940cc050ef57544e61eee9b713ab8955
# good: [ca781c521aab60a3e5f06ca51bfbf608fc74ffb0] Remove interruption check from marching cubes
git bisect good ca781c521aab60a3e5f06ca51bfbf608fc74ffb0
# bad: [e91acd000c44f4668eb1e8a145843c39d0408802] Release resources when appropriate
git bisect bad e91acd000c44f4668eb1e8a145843c39d0408802
# good: [063441db4b56df85b1a225b9487c7bd2017b8cd4] Bind enable/disable toggle of intersecting state meshes to label state meshes
git bisect good 063441db4b56df85b1a225b9487c7bd2017b8cd4
# good: [3e71aef75b033fdaa1456d57b613f9af5011a75b] Pass isEnabled only as an argument for MeshSettingsNode.createTitledPane
git bisect good 3e71aef75b033fdaa1456d57b613f9af5011a75b
# good: [8ed9044483634149f05779ca4591779af3e82bd4] Re-establish mesh export button
git bisect good 8ed9044483634149f05779ca4591779af3e82bd4
# bad: [1be0d52bd55851941b180001bb0367d88042d836] Simplify mesh management
git bisect bad 1be0d52bd55851941b180001bb0367d88042d836
# first bad commit: [1be0d52bd55851941b180001bb0367d88042d836] Simplify mesh management
```

hanslovsky commented:
The issue seems to be [`AdaptiveResolutionMeshManager.meshesAndViewerEnabledListenersInterruptGeneratorMap`](https://github.com/saalfeldlab/paintera/blob/1be0d52bd55851941b180001bb0367d88042d836/src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/managed/adaptive/AdaptiveResolutionMeshManager.kt#L75)
```kotlin
private val meshesAndViewerEnabledListenersInterruptGeneratorMap = mutableMapOf<MeshGenerator<ObjectKey>, ChangeListener<Boolean>>()
````
introduced in 1be0d52bd55851941b180001bb0367d88042d836
Strangely, this works:
```kotlin
private val meshesAndViewerEnabledListenersInterruptGeneratorMap: MutableMap<MeshGenerator<ObjectKey>, ChangeListener<Boolean>> = mutableMapOf()
```

hanslovsky commented:
Selecting all ids is still slow (I am testing on lauritzen block 3 with my quasi-isotropic predictions). It is not finding the ids that is slow (that takes in the order of 0.1 seconds). I will move binding/unbinding settings onto a separate non-UI thread again.

hanslovsky commented:
I fixed updating the meshes on merge/detach actions. A remaining issue is updating the meshes after painting: The label block lookup seems to generate incorrect blocks at lower resolutions but it works fine for the first two scale levels in my CREMI example. After committing the canvas, everything looks fine for all scale levels, so it may be an issue with the `MaskedSource`. I will have to investigate if and what I changed there.

hanslovsky commented:
There may be an issue with synchronization/locking:
```
   java.lang.Thread.State: BLOCKED (on object monitor)
        at org.janelia.saalfeldlab.paintera.meshes.managed.adaptive.AdaptiveResolutionMeshManager.cancelAndUpdate(AdaptiveResolutionMeshManager.kt:229)
        - waiting to lock <0x000000069454b118> (a org.janelia.saalfeldlab.paintera.meshes.managed.adaptive.AdaptiveResolutionMeshManager)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments$managerCancelAndUpdate$1.invalidated(MeshManagerWithAssignmentForSegments.kt:130)
        at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:349)
        at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
        at javafx.beans.property.IntegerPropertyBase.fireValueChangedEvent(IntegerPropertyBase.java:106)
        at javafx.beans.property.IntegerPropertyBase.markInvalid(IntegerPropertyBase.java:113)
        at javafx.beans.property.IntegerPropertyBase.bind(IntegerPropertyBase.java:204)
        at org.janelia.saalfeldlab.paintera.meshes.MeshSettings.bindTo(MeshSettings.kt:187)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments.setupGeneratorState(MeshManagerWithAssignmentForSegments.kt:207)
        - locked <0x000000069454a128> (a org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments.createMeshFor(MeshManagerWithAssignmentForSegments.kt:199)
        - locked <0x000000069454a128> (a org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments.access$createMeshFor(MeshManagerWithAssignmentForSegments.kt:50)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments$setMeshesToSelection$task$1.invoke(MeshManagerWithAssignmentForSegments.kt:178)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments$setMeshesToSelection$task$1.invoke(MeshManagerWithAssignmentForSegments.kt:50)
        at org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithAssignmentForSegments$CancelableTask.run(MeshManagerWithAssignmentForSegments.kt:70)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

```

igorpisarev commented:
@hanslovsky Looks like a deadlock, is there a second thread that owns the lock `<0x000000069454b118>` and perhaps wants to acquire `<0x000000069454a128>`?

hanslovsky commented:
Thanks @igorpisarev 

I have multiple `ExecutorServices` for binding/unbinding (setting up) the mesh generator states because they get set up in part in the `AdaptiveResolutionMeshManager` and in part in the mesh manager that wraps it, e.g. `MeshManagerWithAssignmentForSegments`. I am currently re-writing this to only set up the states inside the `AdaptiveResolutionMeshManager`, which will make threading and locking simpler.

hanslovsky commented:
I will go ahead and merge/squash this into the original `3d-visualize-masks` branch and continue there.

hanslovsky commented:
I will not delete this branch for now so we can keep the commit history for a while. Eventually, this branch should be deleted.
